### ISSUE TITLE
Implement a concept of "Required" properties

### ISSUE THREAD

**Issue by ryanbrandenburg** [dotnet/runtime#29861]
[AB#1130869](https://devdiv.visualstudio.com/10e66e43-9645-4201-b128-0fdc3769cc17/_workitems/edit/1130869)

### API proposal:

```csharp
namespace System.Text.Json.Serialization
{
  [AttributeTargets.Field | AttributeTargets.Property, AllowMultiple=false)]
  public sealed class JsonRequiredAttribute : System.Text.Json.Serialization.JsonAttribute
  {
    public JsonRequiredAttribute() {}
  }
}

namespace System.Text.Json.Serialization.Metadata
{
  public abstract partial class JsonPropertyInfo
  {
    public bool IsRequired { get; set; }
  }
}
```

### Current situation
There's no way to indicate that if an object doesn't have a particular property that serialization should fail.

For example:
```c#
public class Person
{
    [JsonRequired]
    public string Name { get; set; }
    public int Age { get; set; }
}
```
```json
{
    "Age": 46
}
```

Trying to deserialize the above json as a `Person` will succeed, even though it's been indicated that the name attribute is required, and is missing.

### Describe the solution you'd like
System.Text.Json should respect either `System.ComponentModel.DataAnnotations.RequiredAttribute` or a new attribute (likely extending `System.Text.Json.JsonAttribute`) and error when an attempt is made to serialize an object which doesn't have that property.

### Describe alternatives you've considered
Using `System.ComponentModel.DataAnnotations.RequiredAttribute` might be troublesome because it could indicate Requirement in a different context. The alternative is that users implement their own "Required"  functionality (and likely get it wrong or incomplete).

CC @pranavkm who I talked about this with.

**Comment by markfinlabs** [comment]
> .... if `Name` is a non-nullable reference type, wouldn't that indicate that it was required?
> 
> So:
> 
> ```cs
> public class Person
> {
>     public string Name { get; set; }
>     public int Age { get; set; }
> }
> ```
> 
> would fail the deserialize, but:
> 
> ```cs
> public class Person
> {
>     public string? Name { get; set; }
>     public int Age { get; set; }
> }
> ```
> 
> ... would let it succeed.

I agree such behaviour would be very desirable to ensure compiler checks with nullable reference types are not undermined at runtime by incorrect JSON (defeating the purpose of nullable). However this is perhaps best handled via the separate issue https://github.com/dotnet/runtime/issues/1256.

A Required attribute would at least allow a reasonable manual workaround for this in the meantime though.

NB: I have sadly decided to roll back my conversion to System.Text.Json since as now using nullable reference types I don't want to go back to adding lots of nullable run-time checks.

**Comment by jebbinBjss** [comment]
@steveharter 

> Required was discussed but didn't meet the bar for 3.0 in lieu of other features. A workaround is to create a custom converter (pending feature) and override the Read method.

I see that this issue has been considered a nice to have, but I would like to point out that the workaround of creating a custom converter isn't an extensible one, as you would either need to create a custom converter for every class with required properties, or create a generic one that allows you to pass the names of all the required fields to the converter, while also handling the specifics of actually converting the values.

**Comment by Gambero81** [comment]
What about the progress of this issue?

It would be glad to have the JsonRequired attribute used in Newtonsoft.Json implemented in System.Text.Json

**Comment by ijagberg** [comment]
Also interested in a progress update on this issue. The workaround described [here](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-5-0#required-properties) is not viable when you are working with many different classes.

**Comment by steveharter** [comment]
A notification feature was recently implemented that will be in Preview 7: https://github.com/dotnet/runtime/issues/54528

This assumes the Type is owned, as it relies on implementing interfaces directly on the Type.

This approach is much easier than the custom converter approach, plus allows any type of validation logic such as dependent properties (e.g. if "A" is non-null, then "B" must be non-null).

Here's an example that validates during both JsonSerializer.Serialize() and JsonSerializer.Deserialize() to ensure a `FirstName` property is not null.
```cs
public class Person : IJsonOnDeserialized, IJsonOnSerializing
{
    public string FirstName{ get; set; }

    void IJsonOnDeserialized.OnDeserialized() => Validate(); // Call after deserialization
    void IJsonOnSerializing.OnSerializing() => Validate(); // Call before serialization

    private void Validate()
    {
        if (FirstName is null)
        {
            throw new InvalidOperationException("The 'FirstName' property cannot be 'null'.");
        }
    }
}
```

**Comment by jeremyVignelles** [comment]
@steveharter : This example works well for types where an "invalid" value can be reserved, but what about `bool` for example ? How do you say that "The caller must provide a boolean value here" ?
I don't think you can detect that after serialization, and while custom validation after deserialization is probably valuable, it doesn't cover the use case I'm highlighting : How to make sure that a required property is not missing from the json ?

**Comment by layomia** [comment]
@jeremyVignelles would using a nullable `bool` i.e. `bool?` work for you? That way you can determine whether a non-null value was provided.

**Comment by layomia** [comment]
Moving this feature to future given that there's an alternative approach shown in https://github.com/dotnet/runtime/issues/29861#issuecomment-884437381.

**Comment by jeremyVignelles** [comment]
> @jeremyVignelles would using a nullable `bool` i.e. `bool?` work for you? That way you can determine whether a non-null value was provided.

Probably fine by me, I'm in favor of having business models separate from the JSON one.

However, that might not be the case for everyone, and having to change a bool to a Nullable<bool> is a breaking change for the code that uses it.

Moreover, it doesn't let you have a default value for the boolean value if it's not provided, so you need to handle that default value elsewhere.

**Comment by Gambero81** [comment]
The validation need to be done over json and not on model before/after serialization/deserialization.

Newtonsoft.Json does it!

**Comment by HappyPepi** [comment]
We really would like to use System.Text.Json instead of Newtonsoft.Json in our projects but this missing feature stops our adoption as the suggested approaches/workarounds are not suitable for us and Newtonsoft.Json is currently more convenient.

**Comment by Nickprovs** [comment]
@Gambero81 Yeah, Newtonsoft's JsonProperty and JsonRequired attribute are incredibly convenient. They allow the consumer to mark **structs** like DateTime as [JsonRequired] without making it nullable. If the DateTime is not present in the model, it adds an error to the ModelState, and the default value is set for DateTime. 

This is so much more convenient than using the ASP.NET's [Required] attribute, which occurs after model binding. Marking a DateTime with Asp.Net's [Required] attribute will not add an error to the model state because by the time it runs, the model will already have been bound and it will be validating the model object rather than the JSON. And you can't have a null DateTime, so it will be validating the default value. This is a problem for all structs including basic things like bool.

In order to get this to work, you'd have to mark your struct as nullable and deal with all the headaches that creates downstream with code and static analysis tools flagging your usage.

**Comment by eiriktsarpalis** [comment]
I don't believe that `IJsonOnDeserialized` is adequate to encode a "Required" attribute, since typically the intention is to ensure that the underlying JSON contains data for the property in question; not to validate whether the value of the property is `default`.

In Newtonsoft for example, there is the `JsonRequired` attribute which can be used precisely for that reason:
```csharp
using Newtonsoft.Json;

JsonConvert.DeserializeObject<MyClass>(@"{""Value"" : 0}"); // succeeds
JsonConvert.DeserializeObject<MyClass>(@"{}"); // fails with validation error

internal class MyClass
{
    [JsonRequired]
    public int Value { get; set; }
}
```
In the above example, a callback-based approach would not be sufficient, since the hydrated instance would be identical for both JSON inputs.

**Comment by eiriktsarpalis** [comment]
This should likely be tackled in conjunction with #63686 cc @krwq

**Comment by MythicManiac** [comment]
Just want to drop in and mention I've recently run into this, given that `JsonZerializer.Deserialize` returns invalid data if you deserialize a type that has explicitly non-nullable fields on it. If nullability is enabled (as it is by default for newer solutions), the compiler will assume those non-nullable fields to always be non-null, yet the deserializer might return them with null values, thus breaking some static analysis guarantees.

This further propagates into different bugs that would normally be caught by static analysis during compile time (nullable value treated as if it was never null), including even warnings when you do add validation, such as the one below:
![image](https://user-images.githubusercontent.com/8225825/153729842-dd87735f-029b-4e24-afa6-057e411edcf5.png)

In other words, the type-generic variant of `JsonSerializer.Deserialize` method actively breaks the guarantees offered by static analysis.. The only workaround I can think of is to always mark all fields use with json deserializer as nullable, and have a separate process for converting them to "validated" types.

Whether or not validators are something that should be included in the default Deserializer is a smaller problem overall, but if it can't guarantee fields that are typed as non-nullable to actually not be null, maybe it should disallow working with non-nullable types entirely given that it just doesn't work?

**Comment by Hoffs** [comment]
> thus breaking some static analysis guarantees

how do you get around the fact that you have non nullable type (`string`) without any initial value set? `public string FileMd5 { get; set; }` should be generating a warning at property level that your property can actually be null.

**Comment by MythicManiac** [comment]
The same way as how you can declare non-nullable types that are filled in the constructor, meaning it's fine to declare such properties as long as they are populated appropriately during the creation of the object.

That is another good point however, the compiler as it is right now does not seem to issue an error if you attempt to initialize an object without all of it's non-nullable fields set to non-null values during initialization. How related that is to this issue I'm unsure, but it does surprise me how lax the compiler is with nullability checks like this.

**Comment by Hoffs** [comment]
> The same way as how you can declare non-nullable types that are filled in the constructor, meaning it's fine to declare such properties as long as they are populated appropriately during the creation of the object.

I don't follow. If you have object initialization path that ensures that after creation of the instance these properties are not null then that is not a responsibility of static analysis to guess that a property that is said to be non null can be null. Just like setting a value of non nullable property to null just generates a warning on that part where it is setting the value. That won't affect the rest of the analysis because you are breaking the assumption at one point.

> the compiler as it is right now does not seem to issue an error if you attempt to initialize an object without all of it's non-nullable fields set to non-null values during initialization

It does though?

```
public class ClassDomain
{
    [Required]
    public string Name { get; set; } = string.Empty;
    public string NonNullUninitialized { get; set; }
    public int Pass { get; set; }
    public int? Key { get; set; }
}
```

yields warning:

> /Class1.cs(10,19): warning CS8618: Non-nullable property 'NonNullUninitialized' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/home/hoffs/projects/RequestDto/IntegrationTest/IntegrationTest.csproj]

**Comment by MythicManiac** [comment]
@Hoffs I'm confident I have a project with the issue I presented, but I don't have it available here right now and given it's going a bit off topic, let's just leave it as IMO it'd be better to explicitly disallow non-nullable fields for the serializer, or alternatively make it respect the constraints they imply.

**Comment by jeffhandley** [comment]
This feature is being moved out of .NET 7 into the Future milestone. We will continue to follow the C# proposal for required properties as a possible integration.

**Comment by krwq** [comment]
I've internally added support for `JsonPropertyInfo.IsRequired` in https://github.com/dotnet/runtime/pull/72937 since that was - I think we should be able to fix this issue with few lines of product code on top of that + testing. I'll add it back to 7.0 and soon update the API proposal.

**Comment by terrajobst** [comment]
* Looks as proposed

```C#
namespace System.Text.Json.Serialization
{
    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property,
                    AllowMultiple=false)]
    public sealed class JsonRequiredAttribute : JsonAttribute
    {
        public JsonRequiredAttribute();
    }
}

namespace System.Text.Json.Serialization.Metadata
{
    public abstract partial class JsonPropertyInfo
    {
        public bool IsRequired { get; set; }
    }
}
```
